src/error.rs

```diff
@@ -139,3 +139,65 @@ where
 {
     DbErr::Json(s.to_string())
 }
+
+/// An error from unsuccessful SQL query
+#[derive(Error, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum SqlErr {
+    /// error for inserting a record with a key that already exists in the table
+    #[error("Cannot have record with same key")]
+    UniqueConstraintViolation(),
+    /// error for Foreign key is not primary key
+    #[error("Cannot add non-primary key from other table")]
+    ForeignKeyConstraintViolation(),
+}
+
+#[allow(dead_code)]
+impl DbErr {
+    /// converting generic DbErr from mysql to SqlErr
+    #[cfg(any(
+        feature = "sqlx-mysql",
+        feature = "sqlx-postgres",
+        feature = "sqlx-sqlite"
+    ))]
+    fn sql_err<E: sqlx::error::DatabaseError>(&self) -> Option<SqlErr> {
+        if let DbErr::Exec(RuntimeErr::SqlxError(sqlx::Error::Database(e)))
+        | DbErr::Query(RuntimeErr::SqlxError(sqlx::Error::Database(e))) = self
+        {
+            #[cfg(feature = "sqlx-mysql")]
+            {
+                if e.try_downcast_ref::<SqlxMySqlError>().is_some() {
+                    if e.code().unwrap().eq("1062") | e.code().unwrap().eq("1586") {
+                        return Some(SqlErr::UniqueConstraintViolation());
+                    };
+                    if e.code().unwrap().eq("1452") {
+                        return Some(SqlErr::ForeignKeyConstraintViolation());
+                    };
+                }
+            }
```

Comment:
I believe this can be simply as

```suggestion
                #[cfg(feature = "sqlx-mysql")]
                if e.try_downcast_ref::<SqlxMySqlError>().is_some() {
                    if e.code().unwrap().eq("1062") | e.code().unwrap().eq("1586") {
                        return Some(SqlErr::UniqueConstraintViolation());
                    };
                    if e.code().unwrap().eq("1452") {
                        return Some(SqlErr::ForeignKeyConstraintViolation());
                    };
                }
```